cmd/serve_cmd.go

```diff
@@ -32,10 +37,63 @@ func serve(ctx context.Context) {
 	}
 	defer db.Close()
 
-	api := api.NewAPIWithVersion(config, db, utilities.Version)
-
 	addr := net.JoinHostPort(config.API.Host, config.API.Port)
 	logrus.Infof("GoTrue API started on: %s", addr)
 
-	api.ListenAndServe(ctx, addr)
+	a := api.NewAPIWithVersion(config, db, utilities.Version)
+	ah := reloader.NewAtomicHandler(a)
```

Comment:
Oh OK now I get why you need the atomic handler.

I'd maybe simplify this a bit by getting rid of the AtomicHandler and do something simple like:

```suggestion
	currentHandler := atomic.Value(...)
	httpSrv := &http.Server{
		Addr:              addr,
		Handler:           func() {
			currentHandler.handle.ServeHTTP()
		},
		ReadHeaderTimeout: 2 * time.Second, // to mitigate a Slowloris attack
		BaseContext: func(net.Listener) context.Context {
			return baseCtx
		},
	}
```

That way you don't need to create a new type, etc.

Reply:
It might make more sense to create an abstraction over the HTTP server rather than an abstraction over the "handler."

We could reuse that abstraction later on for other things, while reusing the atomic handler is unlikely and thus unnecessary.

Reply:
A handler is the smallest abstraction possible over handling http requests. This is the first primitive you would need in a mulit-tenant environment, as far as future use goes. I could change this into an abstraction you envision, but you'll need to be prescriptive here cause I'm not seeing it. Sorry.

Reply:
Sure this works as well... just offering an alternative that might (but not significantly probably) cut down on code changes.